oracles/oracle.py

```diff
@@ -14,34 +15,39 @@
 
 CHAT_TASKS = {}
 FUNCTION_TASKS = {}
+MAX_CONCURRENT_CHATS = 5
+MAX_CONCURRENT_FUNCTION_CALLS = 5
 
 
-async def _answer_chat(chat: Chat):
+async def _answer_chat(chat: Chat, semaphore: Semaphore):
     try:
-        if chat.response is None:
-            response = await generate_response_use_case.execute(
-                "gpt-4-turbo-preview", chat
+        async with semaphore:
+            print(f"Answering chat {chat.id}", flush=True)
+            if chat.response is None:
+                response = await generate_response_use_case.execute(
+                    "gpt-4-turbo-preview", chat
+                )
+                chat.response = response.chat_completion
+                chat.error_message = response.error
+
+            success = await repository.send_chat_response(chat)
+            print(
+                f"Chat {chat.id} {'' if success else 'not '}"
+                f"replied, tx: {chat.transaction_receipt}",
+                flush=True
             )
-            chat.response = response.chat_completion
-            chat.error_message = response.error
-
-        success = await repository.send_chat_response(chat)
-        print(
-            f"Chat {chat.id} {'' if success else 'not '}replied, tx: {chat.transaction_receipt}",
-            flush=True
-        )
     except Exception as ex:
         print(f"Failed to answer chat {chat.id}, exc: {ex}", flush=True)
 
 
 async def _answer_unanswered_chats():
     while True:
         try:
+            semaphore = asyncio.Semaphore(MAX_CONCURRENT_CHATS)
```

Comment:
Probably move this outside while loop because it will create a new semaphore for each loop run.

Reply:
Actually not.
The while loop is waiting until all current tasks are finished before getting all the new tasks again and creating a new semaphore.
Tested it out as well and makes sense but just in case can move it out side of the while loop.